### ISSUE TITLE
View: Rendering and handling (in mutations) spaces at block boundaries and multiple spaces.

### ISSUE THREAD

**Issue by Reinmar** [ckeditor/ckeditor5-engine#379]
We need to figure out what to do with https://github.com/ckeditor/ckeditor5-design/issues/61

Imagine that your model contains paragraph with `"foo "`. This needs to be rendered as `<p>foo&nbsp;</p>`.

Another case: user is typing at the end of a paragraph – "foo bar". We have "foo", then user presses space. That space must be rendered as `&nbsp;` (we actually get `&nbsp;` in a mutation). Then, two things may happen:

1. If user moved caret somewhere else and back to that place or we rerendered the selection, then browser *may* insert "b" without turning that nbsp into a normal space.
2. Typically (at least Chrome) converts that nbsp into normal space once "b" was typed. I checked that I can move selection around or load such data and the conversion takes place. I'm not sure however that it's going to work for all browsers and in all cases, so it's safe to assume that it won't.

So:

* A space typed at the end of a paragraph should be inserted as a normal space into the model. We may add a feature that <kbd>Alt + Space</kbd> (or whichever keystroke we'll choose from https://en.wikipedia.org/wiki/Non-breaking_space#Keyboard_entry_methods) will insert an `nbsp` to the model. Existence of such a feature shows that the model must always contain clean data.
* Multiple subsequent spaces must be rendered correctly.
* Mutations should rather contain un-quirked text content. So we assume that a space typed at the end of block, even though inserted as an nbsp by the browser should really be a normal space.
* However, the above should not be true for characters which were added as nbsps into the model and rendered as such. DOM converter must not convert them to normal spaces. This may be tricky ;/.

Option – we may for now use the CSS property which I described in ckeditor/ckeditor5-design#61 and do not care about this at all. The editor should work perfectly.

**Comment by pjasiun** [comment]
Makes sense. `DOMConverter`, together with `Renderer` and `MutationObserver` should handle it, so the View will be clear and beautiful. One day. ;)

**Comment by pjasiun** [comment]
### TC 3

 - editor with typing feature,
 - model data: `foo [bar] bom`,
 - press 'x'.

**Expected result:**
`foo x[] bar`

**Actual result:** 
`foo x[]bar`

**Reason:**
What typing is doing is: 
```
deleteContents( selectedContetn );
render();
handleMutations();
render();
```
After the `deleteContents` there is `foo  bom` in the model. The two sibling spaces are rendered to the single space by the browser. Then, when mutation which handle insertion of the "x" character comes, it do the clean-up too and removes unneeded space, so there are two mutations: "remove space" and "insert x".

We should render one of these spaces as `&nbsp;` to prevent it.

**Comment by scofalik** [comment]
Originally submitted in: https://github.com/ckeditor/ckeditor5-typing/issues/30

## TC4

Similarly to **TC1* and *TC2**:

**Steps to reproduce:**
* Write "Hello world"
* Remove "world"

**What happens:**
The word "world" is removed. However, the space that was between "Hello" and "world" is also removed. 

## TC5

**Steps to reproduce:**
* Write "Hello world"
* Remove "Hello"

**What happens:**
Both "Hello" and space between words are removed.

In both cases, space should be converted to `&nbsp;` to ensure correct rendering.

**Comment by Reinmar** [comment]
OK, this issue is getting serious, because it's pretty evident and fails quickly.

To the cases described above I'd also add:

```html
<heading1>Heading 1</heading1>
<heading2>[] </heading2>
<paragraph><$text italic="true">This</$text> is an <$text bold="true">editor</$text> instance.</paragraph>
```

The `heading2` element is rendered with a normal space, so it's invisible.

**Comment by Reinmar** [comment]
Related issue: https://github.com/ckeditor/ckeditor5-engine/issues/606. Defines that spaces must be normalised when converting back to the model. I guess that both issues must be fixed together.

**Comment by scofalik** [comment]
Okay, let's summarize some knowledge that is spread around multiple issues and what we concluded:

1. We want to keep model clean. If user pressed space, we want breakable space in model. If user pressed <kbd>shift + space</kbd> we want `&nbsp;` in the model. Multiple consecutive spaces should be just spaces in the model.

2. We would like to keep view clean aswell. We understand that it's perfect when view is as similar to DOM as it can be, but view is the entry point for some algorithms (mostly data processors) and we don't want to insert quirks into view. I see the need of normalize/replace characters as kind-of-a-quirk.

3. We'll try to make all the changes in the renderer. It would be better if we don't need to use `white-space: pre-wrap;` but we consider this solution as last restort.

4. Upon rendering from view to DOM, we need to:
 - change consecutive spaces to `&nbsp;` as needed,
 - change spaces at boundaries to `&nbsp;`
 - make sure that we are not leaving any unwanted `&nbsp;`s.

5. We need to handle mutations for `space` key event:
 - <kbd>shift + space</kbd> inserts `&nbsp;` into contenteditable, so we do not need to do anything,
 - <kbd>space</kbd> may insert either space or `&nbsp;`. Whenever `&nbsp` is inserted, we need to change it to normal space (and let model -> view -> DOM pipeline handle appropriate changes).

6. We need to handle setting contents (loading, pasting, etc.):
 - whitespaces (or any text) in wrong context should be removed, accordingly to rules set in `Schema`,
 - whitespaces in correct context should be handled, but.. how?

I think that most stuff that has to be done is clear but loading the content is not really, at least not for me. Let's say that somebody initializes editor with such html:
```html
<div id="editor">
    <p>
        myText
    </p>
</div>
```
What about whitespaces inside `<p>`? How do we treat them? I guess we should remove all of them and treat this part of the code as `<p>myText</p>`. However, then, I am afraid we might have a conflict on how we treat whitespaces (see point 5.).

Keep in mind that I am not that familiar with `Renderer` and all code around it. I know what are the main goals and conceptions about it but still feel a bit lost. That's why I am not sure if it will be easy to handle pasting and typing in `Renderer` and related code differently.

**Comment by Reinmar** [comment]
> I see the need of normalize/replace characters as kind-of-a-quirk.

I agree. Especially that it may depend on page styling. DOM renderer is the only bit which should be concerned about that.

> We'll try to make all the changes in the renderer. It would be better if we don't need to use white-space: pre-wrap; but we consider this solution as last restort.

We always tried to not depend on certain page styling and not to force anything. 

> shift + space inserts &nbsp; into contenteditable, so we do not need to do anything,

This may require a separate handler for <kbd>Shift+Space</kbd> because we won't be able to distinguish what was the user intention based on mutations.

> whitespaces in correct context should be handled, but.. how?

They need to be normalised... Either upon conversion (view->model) or by the DOM converter (DOM->view). The latter won't handle spaces if e.g. Markdown DP was used, because then the DOM converter may not be used (MD->view->model).

> What about whitespaces inside <p>? How do we treat them?

All should be removed.

> However, then, I am afraid we might have a conflict on how we treat whitespaces (see point 5.).

Hmmm.... that's true.

**Comment by scofalik** [comment]
~~<kbd>shift + space</kbd> inserts &nbsp; into contenteditable, so we do not need to do anything,~~

This is actually wrong, <kbd>shift + space</kbd> does not insert `&nbsp;` :(

**Comment by scofalik** [comment]
After some digging into the problem here is what I discovered and how I plan to solve this issue:

1. Enhance `DomConverter#viewToDom` so it takes care of converting spaces in view to `&nbsp`s. This deals with point 4. above. I also changed `Renderer#_updateText` because it also takes data from `ViewText#data` and inserts directly into dom text node. Now, it first converts `ViewText` node to dom text node using `DomConverter#viewToDom` so all quirks that will be implemented in `DomConverter` will also be used in `_updateText`.

2. Pasting is now, obviously, not available, so for point 6. I focus on setting initial data. This is done through `DataController` which calls data processor, which then gets the data and converts it to view. We care about `HTML` and `XML` data processors. They both use `DomConverter#domToView` and `DomConverter#viewToDom`. I think its appropriate to modify `domToView`. Nothing else in engine right now uses this and I suppose nothing else in other repos. So it's safe to modify there. I will have to remove all unimportant whitespaces:
 - every white space character should be changed to space (?),
 - double space should be changed to single space,
 - space character at the beginning of block node should be removed - this is actually tricky part. It's tricky because at that point we do not know which view elements are blocks (`ContainerElement`) and which are inline (`AttributeElement`). We know it only after conversion from view to model and back. So I am not really sure how to solve this if we want to avoid putting this logic in view->model conversion (we need to check whether it's safe to put that logic there).

3. Mutations operate directly on model using `Batch` API. The easiest place to modify is directly before `batch.weakInsert()` is called: https://github.com/ckeditor/ckeditor5-typing/blob/master/src/input.js#L234-L238
Here we can just replace in `text` all instances of `&nbsp;` with ` `. Unfortunately it will be harded to implement <kbd>shift + space</kbd> behavior when we go this way. Another solution is to treat it like we did with <kbd>enter</kbd>. It has it's own event and does not rely on mutation. However, I suspect that mutations are used for a reason so I don't know if moving space key handling out of them is good idea. However, AFAICS, when handling mutations, we do not have information about pressed keys. So, AFAICS, it's impossible to check whether user had <kbd>shift</kbd> key pressed. This means that we probably have to manage this key combination like we do with `<kbd>enter</kbd>. We can keep normal <kbd>space</kbd> handling in mutations, though.

**Comment by pjasiun** [comment]
1 and 2 are fine for me. 

But:

> 3 Mutations operate directly on model using Batch API.

No, they do not. They fire only the event which should have view data. For text there was not trasformation, because we did nothing with text. No the text in the event fired from the mutation observer should be "view text".

**Comment by Reinmar** [comment]
Agree with @pjasiun.
